src/looplets/pipelines.jl

```diff
@@ -99,4 +108,45 @@ function (ctx::PipelineVisitor)(node::Shift)
     map(PipelineVisitor(; kwfields(ctx)..., ext = shiftdim(ctx.ext, call(-, node.delta)))(node.body)) do (keys, body)
         return keys => Shift(body, node.delta)
     end
-end
\ No newline at end of file
+end
+
+
+# Collecting End Points of Phase in Pipelines
+Base.@kwdef struct PipelineEndpointsCollector
+    ctx
+    ext
+end
+
+function (ctx::PipelineEndpointsCollector)(node)
+    if istree(node)
+        map(flatten((product(map(ctx, arguments(node))...),))) do phases
+            keys = map(first, phases)
+            bodies = map(last, phases)
+            return reduce(vcat, keys, init=[]) => reduce(vcat, bodies, init=[]) 
+        end
+
+    else
+      return [[] => []]
+    end
+end
+function (ctx::PipelineEndpointsCollector)(node::FinchNode)
+    if node.kind === virtual
+        ctx(node.val)
+    elseif istree(node)
+        map(flatten((product(map(ctx, arguments(node))...),))) do phases
+            keys = map(first, phases)
+            bodies = map(last, phases)
+            return reduce(vcat, keys, init=[]) => reduce(vcat, bodies, init=[]) 
+        end
+    else
+      return [[] => []]
+    end
+end
+
+function (ctx::PipelineEndpointsCollector)(node::Pipeline) 
+  phasestops = map((phase) -> (isnothing(phase.stop(ctx.ctx, ctx.ext)) ? getstop(ctx.ext) : phase.stop(ctx.ctx, ctx.ext)), 
+                   node.phases)
+  return enumerate(append!([simplify(call(-,getstart(ctx.ext), 1), ctx.ctx)], phasestops))
+end
+
+
```

Comment:
This could be done by the PipelineVisitor if we copy the stop of each phase to the start of the next phase on line https://github.com/willow-ahrens/Finch.jl/blob/513d2de6a3c46b83d9af0c2d03676e1a3f6bfe4f/src/looplets/pipelines.jl#L105